Add Travis-CI and Appveyor-CI CMake support (fix #109)

Mizux opened the pull request:
# Introduction
This give us a minimal CI for **CMake-based** build on GNU/Linux (Xenial), osx (xcode10.1) and Windows (VS 2017, VS 2015).

Pro:
* Be able to share log and way to reproduce errors (e.g. #192, #193, #195, #198, #244 ...).
* Allow any user to test their branch(s) in their fork before sending a PR (github workflow)
* Travis-CI and Appveyor are fully integrated to github
* Google already pay for extra worker on Travis-CI (ed only for *google/** projects)
* kokoro jobs logs are not accessible from outside and are **useless** when user fork this repo...

Exhaustive jobs List:
* [x] Travis-CI `os:Xenial` `compiler:gcc5.4` `type:Release` `test:on`
* [x] Travis-CI `os:Xenial` `compiler:gcc5.4` `type:Debug` `test:on`
* [x] Travis-CI `os:Xenial` `compiler:clang7.0` `type:Release` `test:on`
* [x] Travis-CI `os:Xenial` `compiler:clang7.0` `type:Debug` `test:on`
* [x] Travis-CI `os:osx` `compiler:xcode10.1` `type:Release` `test:on` (broken cf #244)
* [x] Travis-CI `os:osx` `compiler:xcode10.1` `type:Debug` `test:on` (broken cf #244)
* [x] Appveyor-CI `img:VS 2017 v15.8`  `type:Release` **`test:off`**
* [x] Appveyor-CI `img:VS 2017 v15.8`  `type:Release` `test:on` (broken cf #192 #193 #198)
* [x] Appveyor-CI `img:VS 2017 v15.8`  `type:Debug` `test:on` (broken cf #192 #193 #198)
* [x] Appveyor-CI `img:VS 2015 v14.0`  `type:Release` **`test:off`**
* [x] Appveyor-CI `img:VS 2015 v14.0`  `type:Release` `test:on` (broken cf #192 #193 #198) 
* [x] Appveyor-CI `img:VS 2015 v14.0`  `type:Debug` `test:on` (broken cf #192 #193 #198) 

note: you can see the current job running here:
Travis-CI on my fork https://travis-ci.org/Mizux/abseil-cpp/branches (look at `cmake` branch)
Appveyor-CI on my fork https://ci.appveyor.com/project/Mizux/abseil-cpp/history  (look at `cmake` branch)

note: for Googler see **CL 221070196**

Mizux committed:
Add Appveyor CMake support (#109)

Mizux committed:
Add Travis-CI Cmake jobs (#109)

- Enable Verbose Makefile on unix

Mizux commented:
Currently Appveyor-CI jobs fail due to issue #193 which is expected.
Travis-CI osx fail in test cf #195.

note: my branch `Mizux:master` is based on this branch (i.e. `Mizux:cmake`) and add the fix for #193.

JonathanDCohen commented on .gitignore:
```diff
@@ -1,2 +1,8 @@
 # Ignore all bazel-* symlinks.
 /bazel-*
+build
+# Ignore vim file
```
These comments don't seem to match with what's underneath them

JonathanDCohen commented on .appveyor.yml:
```diff
@@ -0,0 +1,35 @@
+platform: x64
+
+environment:
+  MSBUILD_FLAGS: /verbosity:minimal /maxcpucount
```
Why verbosity:minimal?  I'd rather have diagnostic so we have maximum information on build failures.

JonathanDCohen commented on .appveyor.yml:
```diff
@@ -0,0 +1,35 @@
+version: '{build}'
+shallow_clone: true
```
Please add a comment that this is only cloning abseil at a specific commit. In general, we should err on the side of too many comments here since very few people on Abseil have much experience with appveyor or Travis.  I'm literally sitting here looking up each line in the documentation, so I'd imagine someone in the future debugging a weird build break might have to do the same thing

Mizux replied:
done

JonathanDCohen replied:
Not sure I see the comment

JonathanDCohen commented on .appveyor.yml:
```diff
@@ -0,0 +1,35 @@
+build_script:
+  - cmake --version
+  - cmake -H. -Bbuild -DABSL_RUN_TESTS=ON -DABSL_USE_GOOGLETEST_HEAD=ON -G "%CMAKE_GENERATOR%"
+  - cmake --build build --config Release --target ALL_BUILD -- %MSBUILD_FLAGS%
```
I'm not sure I see where MSBUILD_FLAGS is set

rongjiecomputer replied:
It is in line 7, will be set before running `build_script`.

JonathanDCohen commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+  include:
+    - os: linux
+      dist: trusty
+      compiler: gcc
```
Can we add clang as well while we're here?

JonathanDCohen replied:
Also, what standard library is this using?

Mizux replied:
1) Currently clang is not working.
  details: clang seems to use gcc system include, I'm currently investigating. Let's start little then add it later...
  May be related to https://github.com/travis-ci/travis-ci/issues/9550

2) For GCC take a look at the log
e.g.
https://travis-ci.org/Mizux/abseil-cpp/jobs/445611256#L27-L31 and https://travis-ci.org/Mizux/abseil-cpp/jobs/445611256#L419-L426

JonathanDCohen replied:
surprisingly, I don't see anything there when I try to find the standard library.  I'd assume that this is using libstdc++, but we should at least know the version.  We also have a major user on clang but with libstdc++ (google internal), so this is something we should know. Anyways, this doesn't seem like a blocker, I'll put it on my todos for a followup. 

As for 1, it looks like there is a suggested workaround.  Again, probably suitable for a follow-up.

Mizux replied:
Concerning standard lib, I don't get it.
I means usually gcc and standard lib are shipped together on GNU/linux distro
So since linux Trusty image is shipped with gcc 4.8.2 you'll get standard lib "4.8.2"
cf https://gcc.gnu.org/onlinedocs/libstdc++/manual/api.html#api.rel_480

JonathanDCohen commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+  - cmake --version
+  - cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DABSL_RUN_TESTS=ON -DABSL_USE_GOOGLETEST_HEAD=ON
+  - cmake --build build --target all
+  - cmake --build build --target test -- CTEST_OUTPUT_ON_FAILURE=1
```
we use set TEST_OUTPUT_ON_FAILURE=1 earlier but pass it in a way which looks like a commandline flag here.  I like the first way better, let's try to be as consistent as possible with the scripts in between CI systems

JonathanDCohen commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+      dist: trusty
+      compiler: gcc
+    - os: osx
+      osx_image: xcode9.4
```
We support xcode 7.3.1 and greater, so can we add another matrix row with xcode7.3?

Mizux replied:
first you should focus on `xcode9.4` and investigate why some tests are not passing on macOS (see #195), then add xcode7.3 if you want.

note: since you are in `abseil/` instead of `google/` you have currently only one osx worker...

JonathanDCohen replied:
Fair enough
@gennadiycivil can we investigate getting more osx workers if needed?

JonathanDCohen commented on .travis.yml:
```diff
@@ -0,0 +1,25 @@
+
+script:
+  - cmake --version
+  - cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DABSL_RUN_TESTS=ON -DABSL_USE_GOOGLETEST_HEAD=ON
```
All else being equal, I'd rather choose Debug over Release, but that's unfortunate we can't do better.  Can we not set a BUILD_TYPE env var in the matrix so we get better config coverage?

ahedberg commented:
I think we decided that we want to investigate hooking this up to our internal mirror before merging it. Does that sound right, @JonathanDCohen?

Mizux committed:
Travis-CI: Fix Trusty-clang jobs by using libc++

On Travis you must use -stdlib=libc++ with clang++

Mizux commented:
Add fix for Trusty/clang on Travis-CI
CI report here: https://github.com/Mizux/abseil-cpp/runs/26547610

JonathanDCohen commented:
@ahedberg The idea we had was to accept this for now and then hook them up to our internal mirror after the fact to allow us to use them with internal development as well.

JonathanDCohen commented:
@Mizux could you port this to an internal change as in our contribution guidelines and we can move forward from there?

Mizux commented:
Ok, I'll do it on Monday, need to learn copybara for export...

Mizux committed:
Travis-CI use VM since container is deprecated

- see: https://docs.travis-ci.com/user/reference/trusty/#container-based-infrastructure

Mizux committed:
Travis-CI: bump xcode 9.4 -> 10.1

see: https://docs.travis-ci.com/user/languages/cpp/#clang-on-os-x

Mizux committed:
Travis-CI: bump dist trusty -> xenial

ref:
- https://docs.travis-ci.com/user/languages/cpp/#gcc-on-linux
- https://docs.travis-ci.com/user/languages/cpp/#clang-on-linux

Mizux commented:
Just rebased on master and add jobs without tests enabled on windows since test compilation failed on Appveyor (VS2017 & VS2015)

Mizux commented:
FYI: it seems that Apple clang provided by xcode10.1 does not have a "leak" sanitizer
```shell
[ 58%] Linking CXX executable absl_disabled_leak_check_test
cd /Users/travis/build/Mizux/abseil-cpp/build/absl/debugging && /usr/local/Cellar/cmake/3.12.3/bin/cmake -E cmake_link_script CMakeFiles/absl_disabled_leak_check_test.dir/link.txt --verbose=1
/Applications/Xcode-10.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++  -O3 -DNDEBUG -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/absl_disabled_leak_check_test.dir/leak_check_fail_test.cc.o  -o absl_disabled_leak_check_test libabsl_internal_leak_check_api_enabled_for_testing.a libabsl_leak_check_disable.a ../base/libabsl_base.a ../../lib/libgmock_main.a -fsanitize=leak ../base/libabsl_dynamic_annotations.a ../base/libabsl_internal_spinlock_wait.a ../../lib/libgmock.a ../../lib/libgtest.a 
clang: error: unsupported option '-fsanitize=leak' for target 'x86_64-apple-darwin17.7.0'
make[2]: *** [absl/debugging/absl_disabled_leak_check_test] Error 1
make[1]: *** [absl/debugging/CMakeFiles/absl_disabled_leak_check_test.dir/all] Error 2
make: *** [all] Error 2
The command "cmake --build build --target all" exited with 2.
```
see: https://travis-ci.org/Mizux/abseil-cpp/jobs/474846599#L1164

so this code is wrong:
https://github.com/abseil/abseil-cpp/blob/7ffbe09f3d85504bd018783bbe1e2c12992fe47c/absl/debugging/CMakeLists.txt#L202-L205

since we match `Clang` **AND** `AppleClang` while we should not

JonathanDCohen commented:
Abseil has decided that we're not going to take on travis and appveyor currently, so I will close this issue.  Please feel free to reopen in the future.
